release-21.1: kvserver: fix deadlock in Store.removeInitializedReplicaRaftMuLocked

erikgrinaker opened the pull request:
Backport 1/1 commits from #65781.
Backport 1/1 commits from #67006.

We should let this bake on master for a little bit, in case removing the channel buffer causes any trouble.

/cc @cockroachdb/release @cockroachdb/kv 

Release justification: Fixes for high-priority or high-severity bugs in existing functionality.

---

When `Store.removeInitializedReplicaRaftMuLocked()` is finishing up
replica removal, it calls `replicaScanner.RemoveReplica()` to remove it
from the queue, which happens via a buffered channel. However, this call
(and thus the channel send) was done while still holding `Store.mu`,
which could deadlock if the channel buffer filled up and the
`replicaScanner` also needed to grab `Store.mu`.

Since there is no need for `Store.mu` to be held when calling
`replicaScanner.RemoveReplica()`, this patch unlocks it first. It also
removes the `replicaScanner.removed` channel buffer, to avoid hiding
these sorts of deadlocks in the first place.

Release note (bug fix): Fixed a deadlock that could occur when many
replicas were rapidly queued for removal.

erikgrinaker committed:
kvserver: fix deadlock in Store.removeInitializedReplicaRaftMuLocked

When `Store.removeInitializedReplicaRaftMuLocked()` is finishing up
replica removal, it calls `replicaScanner.RemoveReplica()` to remove it
from the queue, which happens via a buffered channel. However, this call
(and thus the channel send) was done while still holding `Store.mu`,
which could deadlock if the channel buffer filled up and the
`replicaScanner` also needed to grab `Store.mu`.

Since there is no need for `Store.mu` to be held when calling
`replicaScanner.RemoveReplica()`, this patch unlocks it first. It also
removes the `replicaScanner.removed` channel buffer, to avoid hiding
these sorts of deadlocks in the first place.

Release note (bug fix): Fixed a deadlock that could occur when many
replicas were rapidly queued for removal.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65859)
<!-- Reviewable:end -->

erikgrinaker commented:
@cockroachdb/kv This has been baking on `master` for a month now, should we go ahead with the backport or leave it for 21.2?

erikgrinaker commented:
Nevermind:

```
=== RUN   TestAcceptsUnsplitRanges
    test_log_scope.go:73: test logs captured to: /tmp/logTestAcceptsUnsplitRanges808828805
    test_log_scope.go:74: use -show-logs to present logs inline
==================
WARNING: DATA RACE
Read at 0x00c00012c1d0 by goroutine 71:
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*replicaScanner).RemoveReplica()
      /home/erik/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/scanner.go:183 +0x8c
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).removeInitializedReplicaRaftMuLocked()
      /home/erik/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_remove_replica.go:189 +0x884
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).RemoveReplica()
      /home/erik/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/store_remove_replica.go:43 +0xfb
  github.com/cockroachdb/cockroach/pkg/kv/kvserver.TestAcceptsUnsplitRanges()
      /home/erik/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvserver/queue_test.go:669 +0x56f
  testing.tRunner()
      /home/erik/sdk/go1.16.5/src/testing/testing.go:1193 +0x202
```

Don't know why this hasn't been caught by any nightlies, it's on `master` and fails consistently with `stressrace`.

erikgrinaker committed:
kvserver: fix data race on replicaScanner.stopper

In #65781, a race was introduced on `replicaScanner.stopper`, since it
is set by `Start()` and accessed by `RemoveReplica()` but `Start()` is
called asynchronously.

To avoid introducing a mutex around the stopper or refactoring the store
construction, this adds a hack that writes the stopper directly to
`replicaScanner.stopper` synchronously during `Store.Start()`.

Release note: None

erikgrinaker commented:
Above race has been fixed in #67006 which is now included here.
